File: leetcode/python/1356. 根据数字二进制下 1 的数目排序.py
#!/bin/bash python
"""
1356. 根据数字二进制下 1 的数目排序

两种思路：
一是直接借助sort()直接给定自拟的排序接口
二是自行编写自定义的快排
"""

class Solution(object):
    def getOnes(self, n):
        """
        获取数字n中二进制位为1的数目
        """
        ret = 0
        while n > 0:
            ret += 1
            n &= n - 1
        return ret

    def sortByBits(self, arr):
        """
        :type arr: List[int]
        :rtype: List[int]
        """
        if arr == []:
            return []
        
        ret = sorted(arr, key=lambda x : (self.getOnes(x), x))
        return ret
	
	# ===== 第二种方法 =====
    def partition(self, vec, ones, l, h):
        """
        快排分区函数
        """
        start = l - 1
        for i in range(l, h):
            if ones[i] < ones[h] or (ones[i] == ones[h] and vec[i] < vec[h]):
                start += 1
                ones[i], ones[start] = ones[start], ones[i]
                vec[i], vec[start] = vec[start], vec[i]
        start += 1
        ones[h], ones[start] = ones[start], ones[h]
        vec[h], vec[start] = vec[start], vec[h]
        return start

    def _quickSort(self, vec, ones, l, h):
        """
        快排内部递归函数
        """
        if l >= h:
            return
        m = self.partition(vec, ones, l, h)
        self._quickSort(vec, ones, l, m - 1)
        self._quickSort(vec, ones, m + 1, h)
        return

    def sortByBits(self, arr):
        """
        :type arr: List[int]
        :rtype: List[int]
        """
        if arr == []:
            return []
			
        # 生成对应的1的个数矩阵
        ones = []
        for num in arr:
            cnt = 0
            while num:
                cnt += 1
                num &= num - 1
            ones.append(cnt)

        # 调用快排
        self._quickSort(arr, ones, 0, len(arr) - 1)
        return arr

File: leetcode/cpp/1356. 根据数字二进制下 1 的数目排序.cpp
class Solution {
public:
    // 获取一个整数的二进制位中1的个数
    int getOnes(int n) {
        int ret = 0;
        while (n != 0) {
            ret += 1;
            n &= n - 1;
        }
        return ret;
    }
    
    // 自定义的排序函数
    static bool compare(std::pair<int, int> a, std::pair<int, int> b) {
        return a.second < b.second || (a.second == b.second && a.first < b.first);
    }

    vector<int> sortByBits(vector<int>& arr) {
        if (arr.empty()) {
            return arr;
        }

        // 生成pair并排序
        std::vector<std::pair<int, int>> tmp;
        for (auto & i : arr) {
            tmp.push_back(std::make_pair(i, getOnes(i)));
        }
        sort(tmp.begin(), tmp.end(), compare);

        // 生成新的arr
        std::vector<int> ret;
        for (auto & i : tmp) {
            ret.push_back(i.first);
        }
        return ret;
    }

	// ===== 另一种方法 =====
	    // 获取一个整数的二进制位中1的个数
    int getOnes(int n) {
        int ret = 0;
        while (n != 0) {
            ret += 1;
            n &= n - 1;
        }
        return ret;
    }

    // 交换vec两个元素的函数
    void swap(vector<int>& vec, int l, int h) {
        int tmp = vec[l];
        vec[l] = vec[h];
        vec[h] = tmp;
    }

    // 快排分区函数
    int partition(vector<int>& vec, vector<int>& ones, int l, int h) {
        int start = l - 1;
        for (int i = l; i < h; i++) {
            if (ones[i] < ones[h] || (ones[i] == ones[h] && vec[i] < vec[h])) {
                start += 1;
                swap(vec, i, start);
                swap(ones, i, start);
            }
        }
        start += 1;
        swap(vec, h, start);
        swap(ones, h, start);
        return start;
    }

    // 快排内部递归函数
    void _quickSort(vector<int>& vec, vector<int>& ones, int l, int h) {
        if (l >= h) {
            return;
        }
        int m = partition(vec, ones, l, h);
        _quickSort(vec, ones, l, m - 1);
        _quickSort(vec, ones, m + 1, h);
        return;
    }
    
    vector<int> sortByBits(vector<int>& arr) {
        if (arr.empty()) {
            return arr;
        }

        // 生成对应1的个数的数组
        std::vector<int> ones;
        for (auto & i : arr) {
            ones.push_back(getOnes(i));
        }

        // 快排内部递归函数
        _quickSort(arr, ones, 0, arr.size() - 1);
        return arr;
    }
};